static/app/views/dashboards/widgetCard/autoSizedText.tsx

```diff
@@ -0,0 +1,177 @@
+import {useLayoutEffect, useRef, useState, useTransition} from 'react';
+import styled from '@emotion/styled';
+import {useResizeObserver} from '@react-aria/utils';
+import debounce from 'lodash/debounce';
+
+interface Props {
+  children: React.ReactNode;
+  maxFontSize: number;
+  minFontSize: number;
+  calculationCountLimit?: number;
+}
+
+export function AutoSizedText({
+  children,
+  minFontSize,
+  maxFontSize,
+  calculationCountLimit = DEFAULT_CALCULATION_COUNT_LIMIT,
+}: Props) {
+  const [_isPending, startTransition] = useTransition();
+
+  const [parentHeight, setParentHeight] = useState<number | null>(null);
+  const [parentWidth, setParentWidth] = useState<number | null>(null);
+
+  const [fontSize, setFontSize] = useState<number>((maxFontSize + minFontSize) / 2);
+  const [fontSizeLowerBound, setFontSizeLowerBound] = useState<number>(minFontSize);
+  const [fontSizeUpperBound, setFontSizeUpperBound] = useState<number>(maxFontSize);
+
+  const [calculationCount, setCalculationCount] = useState<number>(0);
+
+  const parentRef = useRef<HTMLDivElement>(null);
+  const childRef = useRef<HTMLDivElement>(null);
+
+  const onResize = debounce(() => {
```

Comment:
I cant find it in the spec, but iirc resize observers might already be reasonably throttled by the browser. In any case, I think the approach here is suboptimal because you are re-querying the width and height of the element inside the callback, which causes layout thrashing. The resize observer callback arg should provide the content and bounding box rects which you can just read without having to re-query the dom (this way you avoid layout thrashing which is the most expensive part of the calculation). 

Imo the react aria useResizeObserver hook is poorly implemented and we shouldnt use it for this exact reason